src/maxtree.jl

```diff
@@ -0,0 +1,859 @@
+Base.ndims(maxtree::MaxTree) = ndims(typeof(maxtree))
+Base.length(maxtree::MaxTree) = length(maxtree.parentindices)
+Base.size(maxtree::MaxTree) = size(maxtree.parentindices)
+Base.:(==)(a::MaxTree, b::MaxTree) =
+    (ndims(a) == ndims(b)) && (a.rev == b.rev) &&
+    (a.parentindices == b.parentindices) &&
+    (a.traverse == b.traverse)
+
+"""
+    root_index(maxtree::MaxTree) -> Int
+
+Linear index of the root pixel.
+"""
+root_index(maxtree::MaxTree) = maxtree.traverse[1]
+
+# Gets the path to the root of the max-tree (defined by `ancestors`) starting from
+# `node`. The indices of the nodes along the path are stored into `path`.
+# Returns the index of the root
+@inline function rootpath!(path::Vector{Int},
+                           ancestors::AbstractArray{<:Integer},
+                           node::Integer)
+    empty!(path)
+    p = node
+    @label next
+    push!(path, p)
+    @inbounds q = ancestors[p]
+    if p != q
+        p = q
+        @goto next
+    end
+    return p
+end
+
+# Corrects `maxtree.parentindices`, so that the parent of every pixel
+# is a canonical node. I.e. for each connected component at a level l,
+# all pixels point to the same representative, which in turn points to the
+# representative pixel at the next level.
+#
+# A canonical node ``p`` is either a root of the max-tree, or
+# ``image(parent(p)) != image(p)``
+function canonize!(maxtree::MaxTree, image::GenericGrayImage)
+    @assert size(maxtree) == size(image)
+    parentindices = maxtree.parentindices
+    @inbounds for p in maxtree.traverse
+        q = parentindices[p]
+        if image[q] == image[parentindices[q]]
+            parentindices[p] = parentindices[q]
+        end
+    end
+    return maxtree
+end
+
+# Checks whether the `pixel` + `offset` is still a valid pixel
+# contained inside the image area defined by `axes`.
+isvalid_offset(offset::CartesianIndex{N}, pixel::CartesianIndex{N},
+               axes::NTuple{N}) where N =
+    Base.checkbounds_indices(Bool, axes, (pixel + offset,))
+
+# generates a vector of offsets to the CartesianIndex that defines
+# the neighborhood of the pixel in an N-dimensional image.
+# `connectivity` is the maximal number of orthogonal steps
+# (+1/-1 to a single dimension) that are required to reach a neighbor.
+# Note that the corresponding offsets of linear indices would be sorted in ascending order
+# ensuring efficient memory access pattern.
+function neighbor_cartesian_offsets(::Type{CartesianIndex{N}}, connectivity::Integer) where N
+    (connectivity >= 1) || throw(ArgumentError("connectivity should be positive integer"))
+    return [ci for ci in CartesianIndices(ntuple(_ -> -1:1, N))
+            if 1 <= sum(abs, Tuple(ci)) <= connectivity]
+end
+
+# convert offsets of cartesian indices into linear offsets of the given array
+linear_offsets(offsets::AbstractVector{<:CartesianIndex},
+               arr::AbstractArray) =
+    dot.(Tuple.(offsets), Ref(strides(arr)))
```

Comment:
I should have noticed this earlier; it will be problematic. But it's a lengthier discussion so I'll post it in the main thread.